(DOCS-4352) Update documentation for creating an integration (#13280)
DataDog/integrations-core: docs/dev/new_check_howto.md

---

> title: Create an Agent integration

Author: I'm going to come back and add some further reading pages once we're done with the marketplace docs

  - /guides/new_integration/
---

This guide provides instructions for creating a Datadog Agent integration in the integrations-extras repo. For information on why you'd want to create an Agent integration, see [Creating your own solution][1].

## Prerequisites

> The datadog Agent development tools require:

Reviewer: Is this line in reference to [Developer Toolkit](https://docs.datadoghq.com/developers/integrations/new_check_howto/?tab=configurationtemplate#developer-toolkit) or the required development tools to write an agent integration? 

```suggestion
The required Datadog Agent integration development tools:
```

Author: This is in reference to the required development tools to write an Agent integration 👍

- Python 3.8+.
- [Docker][2] (to run the full test suite)

### Install Python

> Many operating systems come with a pre-installed version of Python. However, the version of Python installed by default might be older than the version used in the Agent, and might lack required tools and dependencies. To ensure that you have everything you need to get an integration running. Install a dedicated Python interpreter.

Reviewer: ```suggestion
Many operating systems come with a pre-installed version of Python. However, the version of Python installed by default might be older than the version used in the Agent, and might lack required tools and dependencies. To ensure that you have everything you need to get an integration running, install a dedicated Python interpreter.
```

Some options for installing Python on your operating system:
- Follow the [official Python documentation][3] to download and install the Python interpreter.
- Use a Python version manager like [pyenv][4].

On most operating systems, Python versions after 3.3 come with a pre-installed version manager called `venv`, which is used in this guide. Debian/Ubuntu installations do not come packaged with `venv`. You can install it by running `sudo apt-get install python3-venv`.

## Set up your development environment

1. Create the `dd` directory and clone the [`integrations-extras` repo][5].

   **Note**: The Datadog development toolkit expects you to work in the `$HOME/dd/` directory. This is not mandatory, but working in a different directory necessitates more configuration steps.

>    To create the `dd` directory and clone the `datadog-extras` repo:

Reviewer: ```suggestion
   To create the `dd` directory and clone the `integrations-extras` repo:
```

                self.service_check('awesome.search', self.WARNING)
```

To learn more about the base Python class, see [Anatomy of a Python Check][9].

## Write tests

There are two basic types of tests:
- Unit tests for specific functionality.
- Integration tests that execute the `check` method and verify proper metrics collection.

Tests are required if you want your integration to be included in `integrations-extras`.

> **Note**: [pytest][10] and [tox][11] are used to run the tests.

Reviewer 2: FYI, we are switching from `tox` to [`hatch`](https://github.com/pypa/hatch). This is a work in progress in `integrations-core` and we plan to migrate extras integration after.

Author: If the tool is creating integrations that use hatch, I think it make sense to update this now 👍

> ### Write a unit test

Reviewer 3: ```suggestion
### Write an unit test
```
?

    c.check({'url': 'http://foobar', 'search_string': 'foo'})
```

`pytest` has the concept of markers that can be used to group tests into categories. Notice that `test_config` is marked as a `unit` test.

The scaffolding is set up to run all tests located in `awesome/tests`. Run the tests:

```bash
ddev test awesome
```

### Write an integration test

The unit test above doesn't check the collection logic. To test the logic, you need an integration test.

#### Create an environment for the integration test

>  The toolchain uses `docker` to spin up an Nginx container and let the check retrieve the welcome page. Create a compose file at `awesome/tests/docker-compose.yml` with the following contents:

Reviewer: Toolchain? 

```suggestion
 The toolkit uses `docker` to spin up an Nginx container and let the check retrieve the welcome page. Create a docker-compose file at `awesome/tests/docker-compose.yml` with the following contents:
```

{{% /tab %}}
{{% tab "Manifest" %}}

> The `awesome/manifest.json` for the Awesome Service Check:

Reviewer: Let's include a link to the manifest.json documentation: https://docs.datadoghq.com/developers/integrations/check_references/#manifest-file

Reviewer: We should also update the below manifest.json example to use a manifest v2 example

Author: Good call! I'll work on adding these

Author: @fanny-jiang Should we also add a line that shows users how to migrate their manifest to the newer version? (`ddev meta manifest migrate awesome 2.0.0`) When I used the tools to generate an integration, it defaulted to the V1 manifest.

Reviewer: I believe manifest version 2 is the default for creating new integrations. Do you have an older version of ddev installed by any chance?

Reviewer: Since this guide is mainly for users creating a new integration, I don't think we need a line showing how to migrate to the newer manifest version.

Author: Great! I agree we don't need to add it if it's included by default in the newer versions.

### What's in the wheel?

> The wheel contains only the files necessary for the functioning of the integration itself. This includes the Check, the configuration example file, and some artifacts generated during the build of the wheel. All of the other elements, including the metadata files are not meant to be contained within the wheel. These latter elements are used elsewhere by the greater Datadog platform and eco-system.

Reviewer: ```suggestion
The wheel contains only the files necessary for the functioning of the integration itself. This includes the Check, the configuration example file, and some artifacts generated during the build of the wheel. All of the other elements, including the metadata files are not meant to be contained within the wheel. These latter elements are used elsewhere by the greater Datadog platform and ecosystem.
```

{{% /tab %}}
{{< /tabs >}}

## Build the wheel

The `pyproject.toml` file provides the metadata that is used to package and build the wheel. The wheel contains the files necessary for the functioning of the integration itself, which includes the Check, configuration example file, and artifacts generated during the build of the wheel.

All additional elements, including the metadata files, are not meant to be contained within the wheel, and are used elsewhere by the Datadog platform and ecosystem. To learn more about Python packaging, see [Packaging Python Projects][17].

Once your `pyproject.toml` is ready, create a wheel:

- (Recommended) With the `ddev` tooling: `ddev release build <INTEGRATION_NAME>`.
- Without the `ddev` tooling: `cd <INTEGRATION_DIR> && pip wheel . --no-deps --wheel-dir dist`.

## Install the wheel

> The wheel is installed using the Agent `integration` command, available in [Agent v6.10.0 ot lsyrt][18]. Depending on your environment, you may need to execute this command as a specific user or with specific privileges:

Reviewer: ```suggestion
The wheel is installed using the Agent `integration` command, available in [Agent v6.10.0 and up][18]. Depending on your environment, you may need to execute this command as a specific user or with specific privileges:
```

sudo datadog-agent integration install -w /path/to/wheel.whl
```

> **Windows Powershell** (Ensure that your shell session has _administrator_ privileges):

Reviewer: ```suggestion
**Windows PowerShell** (Ensure that your shell session has _administrator_ privileges):
```

Author: hehe, yeah I noticed this one and pushed up a fix 😅

**Windows** (Ensure that your shell session has _administrator_ privileges):

<details>
  <summary>Agent <code>v6.11</code> or earlier</summary>
  
  ```ps
  "C:\Program Files\Datadog\Datadog Agent\embedded\agent.exe" integration install -w /path/to/wheel.whl
  ```

</details>

<details open>
  <summary>Agent<code>v6.12</code> or later</summary>

  ```ps
  "C:\Program Files\Datadog\Datadog Agent\bin\agent.exe" integration install -w /path/to/wheel.whl
  ```

> </details>

Reviewer: I think the Windows PowerShell commands need a leading `&` character:

  ```ps
  & "C:\Program Files\Datadog\Datadog Agent\embedded\agent.exe" integration install -w /path/to/wheel.whl
  ```
https://docs.datadoghq.com/agent/basic_agent_usage/windows/?tab=commandline#agent-commands

Also, let's indicate that the commands are for Windows PowerShell

"C:\Program Files\Datadog\Datadog Agent\bin\agent.exe" integration install -w /path/to/wheel.whl
```

## Integration publishing checklist

After you complete your integration, refer back to this list to make sure you've got everything you need.

> To consider an Agent-based integration complete, and thus ready to be included in the core repository and bundled with the Agent package, several prerequisites must be met:

Reviewer 2: I'm not sure to understand this. Integrations in -extras are not bundled with the agent and not included in the integrations-core repository. Customers will need to install the integration manually and they are not officially maintained by Datadog.

Author: Good call, removing this paragraph

- A `README.md` file with the correct format and contents
- A battery of tests verifying metrics collection
- A `metadata.csv` file listing all of the collected metrics
- A complete `manifest.json` file

> - If the integration collects Service Checks, the `service_checks.json` must be complete as well

Reviewer 2: Should we mention that the CI will also perform some validations and it should be green?
